modules/assim.sequential/R/Analysis_sda_block.R

```diff
@@ -0,0 +1,549 @@
+         Y = Y,
+         R = R))
+}
+
+##' @title build.block.xy
+##' @name  build.block.xy
+##' @author Dongchen Zhang
+##' 
+##' @param settings  pecan standard multi-site settings list.  
+##' @param block.list.all List contains nt empty sub-elements.
+##' @param X A matrix contains ensemble forecasts.
+##' @param obs.mean List of dataframe of observation means, named with observation datetime.
+##' @param obs.cov   List of covariance matrices of state variables , named with observation datetime.
+##' @param t time point.
+##' @details This function will add data and constants into each block that are needed for the MCMC sampling.
+##'  
+##' @description This function split long vector and covariance matrix into blocks corresponding to the localization.
+##' 
+##' @return It returns the `build.block.xy` object with data and constants filled in.
+build.block.xy <- function(settings, block.list.all, X, obs.mean, obs.cov, t) {
+  #set q.type from settings.
+  if (settings$state.data.assimilation$q.type == "vector") {
+    q.type <- 1
+  } else if (settings$state.data.assimilation$q.type == "wishart") {
+    q.type <- 2
+  }
+  
+  #grab basic arguments based on X.
+  site.ids <- unique(attributes(X)$Site)
+  var.names <- unique(attributes(X)$dimnames[[2]])
+  mu.f <- colMeans(X)
+  Pf <- stats::cov(X)
+  diag(Pf)[which(diag(Pf)==0)] <- min(diag(Pf)[which(diag(Pf) != 0)])/5 #fixing det(Pf)==0
+  
+  #distance calculations and localization
+  if (!is.null(settings$state.data.assimilation$Localization.FUN)) {
+    Localization.FUN <- get(settings$state.data.assimilation$Localization.FUN)
+    site.locs <- settings$run %>% 
+      purrr::map('site') %>% 
+      purrr::map_dfr(~c(.x[['lon']],.x[['lat']]) %>% as.numeric)%>% 
+      t %>%
+      `colnames<-`(c("Lon","Lat")) %>%
+      `rownames<-`(site.ids)
+    #Finding the distance between the sites
+    dis.matrix <- sp::spDists(site.locs, longlat = TRUE)
+    #turn that into a blocked matrix format
+    blocked.dis <- block_matrix(dis.matrix %>% as.numeric(), rep(length(var.names), length(site.ids)))
+    Pf <- Localization.FUN(Pf, blocked.dis, settings$state.data.assimilation$scalef %>% as.numeric())
+  }
+  
+  #Handle observation
+  #if we do free run or the current obs.mean are all NULL.
+  if (as.logical(settings$state.data.assimilation$free.run) | all(is.null(unlist(obs.mean[[t]])))) {
+    obs.mean[[t]] <- vector("list", length(site.ids)) %>% `names<-`(site.ids)
+    obs.cov[[t]] <- vector("list", length(site.ids)) %>% `names<-`(site.ids)
+    H <- list(ind = seq_along(rep(var.names, length(site.ids))))
+    Y <- rep(NA, length(H$ind))
+    R <- diag(1, length(H$ind))
+  } else {
+    Obs.cons <- Construct.R(site.ids, var.names, obs.mean[[t]], obs.cov[[t]])
+    Y <- Obs.cons$Y
+    R <- Obs.cons$R
+    if (length(Y) > 1) {
+      PEcAn.logger::logger.info("The zero variances in R and Pf is being replaced by half and one fifth of the minimum variance in those matrices respectively.")
```

Comment:
I'd split this up a bit -- first calculate which(diag(R)==0) and then if it's non-zero you should issue a warning or error or maybe even a severe(not an info) including which rows are causing the issue within the message, and then set R.  Also, the message here mentions Pf but there are not Pf calculations here. Outside of a free run, if a user is passing in data with zero observation error that's a mistake on their part, and one that they should really catch quickly as it will really mess up their SDA.

Reply:
1. Fixed.
2. Fixed.